Awful.apk/src/main/java/com/ferg/awfulapp/AwfulFragment.kt

```diff
@@ -0,0 +1,442 @@
+import com.ferg.awfulapp.widget.ProbationBar
+import com.orangegangsters.github.swipyrefreshlayout.library.SwipyRefreshLayout
+import com.orangegangsters.github.swipyrefreshlayout.library.SwipyRefreshLayoutDirection.BOTH
+import com.orangegangsters.github.swipyrefreshlayout.library.SwipyRefreshLayoutDirection.TOP
+import timber.log.Timber
+
+abstract class AwfulFragment : Fragment(), ProgressListener, AwfulPreferences.AwfulPreferenceUpdate {
+    protected var TAG = "AwfulFragment"
+
+    lateinit protected var mPrefs: AwfulPreferences
+    private var mProgressBar: AwfulProgressBar? = null
+    private var probationBar: ProbationBar? = null
+
+    protected var mSRL: SwipyRefreshLayout? = null
+
+    var progressPercent = 100
+
+    protected val mHandler = Handler()
+
+
+    val awfulActivity: AwfulActivity?
+        get() = activity as AwfulActivity?
+
+    protected val awfulApplication: AwfulApplication?
+        get() = awfulActivity?.application as AwfulApplication?
+
+    /** Get this fragment's display title  */
+    /**
+     * Set the actionbar's title.
+     * @param title The text to set as the title
+     */
+    // TODO: fix race condition in ForumDisplayFragment and ThreadDisplayFragment - both restart their loaders in onResume,
+    // both of those set the actionbar title - even in phone mode where only one is visible. Whichever loads last sets the actionbar text
+    @CallSuper
+    open fun setTitle(title: String) {
+        awfulActivity?.let {
+            Timber.d("setTitle: setting for %s", this.javaClass.simpleName)
+            it.setActionbarTitle(title, this)
+        }
+    }
+    abstract fun getTitle() : String?
+
+    override fun onAttach(context: Context?) {
+        super.onAttach(context)
+        if (activity !is AwfulActivity) {
+            throw IllegalStateException("AwfulFragment - parent activity must extend AwfulActivity!")
+        }
+        mPrefs = AwfulPreferences.getInstance(context, this)
+    }
+
+    protected fun inflateView(resId: Int, container: ViewGroup?, inflater: LayoutInflater): View {
+        val v = inflater.inflate(resId, container, false)
+        mProgressBar = v.findViewById(R.id.progress_bar)
+
+        // set up the probation bar, if we have one - use this ID when adding to a layout!
+        probationBar = v.findViewById(R.id.probation_bar)
+        probationBar?.setListener { goToLeperColony() }
+
+        return v
+    }
+
+    override fun onActivityCreated(aSavedState: Bundle?) {
+        super.onActivityCreated(aSavedState)
+        onPreferenceChange(mPrefs, null)
+    }
+
+    override fun onDestroy() {
+        super.onDestroy()
+        this.cancelNetworkRequests()
+        mHandler.removeCallbacksAndMessages(null)
+        mPrefs.unregisterCallback(this)
+    }
+
+    override fun onDetach() {
+        super.onDetach()
+        this.cancelNetworkRequests()
+        mHandler.removeCallbacksAndMessages(null)
+    }
+
+    protected fun displayForumIndex() {
```

Comment:
This is for functions that return a value - in this case it would make `displayForumIndex` return the value of `context.displayForumIndex()`. Which is `void` (or `Unit`) but it still feels a little weird to do. Unless I'm mistaken?

Reply:
Everything's an expression, everything returns a value (which might be ``Unit``) ;) There's nothing weird about it in functional programming, the braces just make a block that returns ``Unit`` when it finishes, so it's the same signature and everything - just more concise on a single line!
(It's also the [recommended practice](https://kotlinlang.org/docs/reference/coding-conventions.html#function-formatting) - they're returning an ``Int`` in that example but like I said, there's no semantic difference between an expression that returns an ``Int`` and one that returns a ``Unit``)

In this code here there's a slight difference, in that the block returns ``Unit`` but the single-expression version can return ``null``, so the actual type is ``Unit?``. I think that's a quirk of Kotlin's type system and I don't think it really matters, the meaning of returning ``Unit`` doesn't really change if it's null - it still represents "no value", y'know?

Reply:
Oh as far as removing it goes... personally I'd leave it for now, some of them are simple calls, others are wrappers that add some magic numbers, some create an intent... I think it's better to keep it all together in one place for now, it'll make it easier to fix later